spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/CriticalPathRetryingKafkaStreamsDelegate.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2020-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder.kafka.streams;
+
+import java.util.concurrent.Callable;
+
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.KStream;
+
+import org.springframework.boot.autoconfigure.kafka.KafkaProperties;
+import org.springframework.cloud.stream.binder.kafka.streams.properties.KafkaStreamsBinderConfigurationProperties;
+import org.springframework.lang.Nullable;
+import org.springframework.retry.RetryPolicy;
+import org.springframework.retry.backoff.FixedBackOffPolicy;
+import org.springframework.retry.policy.SimpleRetryPolicy;
+import org.springframework.retry.support.RetryTemplate;
+
+/**
+ * This class is part of the public API of Kafka Streams binder.
+ * <p>
+ * It provides a way for Kafka Streams applications to retry certain parts of the logic that are in the critical path.
+ * Applications that need to make external calls during runtime such as a relational database lookup, REST endpoint
+ * invocation etc. are much more susceptible to fail for wide array of reasons. When those calls throw an exception,
+ * the Kafka Streams processor will fail and shuts itself down, thus those calls stand in the critical path of the
+ * application. It is often handy, if the application could retry the failing operation in the event of a network failure,
+ * bad connection or a remote service is temporarily unavailable etc. This API essentially allows the application to
+ * wrap the critical path code in a {@link Callable} implementation and retry it based on binder level retry properties.
+ * This will use Spring's {@link RetryTemplate} under the hood to invoke the Callable implementation.
+ * <p>
+ * It supports both terminal ({@link org.apache.kafka.streams.processor.Processor}) and non-terminal
+ * ({@link org.apache.kafka.streams.kstream.Transformer} operations to be wrapped inside the {@link Callable} implementation.
+ * <p>
+ * See the relevant API methods for more information.
+ *
+ * @param <K> Key type
+ * @param <V> Value type
+ * @author Soby Chacko
+ * @since 3.0.9
+ */
+public class CriticalPathRetryingKafkaStreamsDelegate<K, V> {
```

Comment:
It's not at all clear what this simple wrapper offers when you can use a `RetryTemplate` directly, with a much richer API - different retry policies, different back-off policies, retry context, recovery strategies to be called when retries are exhausted.

Reply:
I was just thinking of this as a starting point. We can add the capability to send to some sort of DLT in the future if the users want to send the failing k/v for later retrieval, but continue processing in the main processor.

Reply:
I have been recommending the usage of the `RetryTemplate` directly before, but somehow it never caught on as a usage pattern.